crates/extensions/tedge_mqtt_ext/src/lib.rs

```diff
@@ -0,0 +1,207 @@
+    //     let (sender, receiver) = channel(10);
+    //     self.subscriber_addresses.push((topics, sender.into()));
+    //     (self.publish_channel.0.clone(), receiver)
+    // }
+
+    pub fn add_client(
+        &mut self,
+        subscriptions: TopicFilter,
+        peer_sender: DynSender<MqttMessage>,
+    ) -> Result<DynSender<MqttMessage>, LinkError> {
+        self.subscriber_addresses.push((subscriptions, peer_sender));
+        Ok(self.publish_channel.0.clone().into())
+    }
+
+    /// Add a new client, returning a message box to pub/sub messages over MQTT
+    pub fn new_client(
+        &mut self,
+        client_name: &str,
+        subscriptions: TopicFilter,
+    ) -> SimpleMessageBox<MqttMessage, MqttMessage> {
+        let (sub_message_sender, sub_message_receiver) = mpsc::channel(16);
+        let pub_message_sender = self
+            .add_client(subscriptions, sub_message_sender.into())
+            .unwrap();
+
+        SimpleMessageBox::new(
+            client_name.to_string(),
+            sub_message_receiver,
+            pub_message_sender,
+        )
+    }
+
+    /// FIXME this method should not be async
+    pub(crate) async fn build(self) -> (MqttActor, MqttMessageBox) {
+        let mut combined_topic_filter = TopicFilter::empty();
+        for (topic_filter, _) in self.subscriber_addresses.iter() {
+            combined_topic_filter.add_all(topic_filter.to_owned());
+        }
+        let mqtt_config = self.mqtt_config.with_subscriptions(combined_topic_filter);
+        let mqtt_message_box = MqttMessageBox::new(
+            mqtt_config,
+            self.publish_channel.1,
+            self.subscriber_addresses,
+        )
+        .await
+        .unwrap(); // Convert MqttError to RuntimeError
+
+        let mqtt_actor = MqttActor;
+        (mqtt_actor, mqtt_message_box)
+    }
+}
+
+#[async_trait]
+impl ActorBuilder for MqttActorBuilder {
+    async fn spawn(self, runtime: &mut RuntimeHandle) -> Result<(), RuntimeError> {
+        let (mqtt_actor, mqtt_message_box) = self.build().await;
+
+        runtime.run(mqtt_actor, mqtt_message_box).await?;
+        Ok(())
+    }
+}
+
+struct MqttMessageBox {
+    mqtt_client: mqtt_channel::Connection,
+    peer_receiver: Receiver<MqttMessage>,
+    peer_senders: Vec<(TopicFilter, DynSender<MqttMessage>)>,
+}
+
+impl MqttMessageBox {
+    async fn new(
+        mqtt_config: mqtt_channel::Config,
+        peer_receiver: Receiver<MqttMessage>,
+        peer_senders: Vec<(TopicFilter, DynSender<MqttMessage>)>,
+    ) -> Result<Self, MqttError> {
+        let mqtt_client = mqtt_channel::Connection::new(&mqtt_config).await?;
+        Ok(MqttMessageBox {
+            mqtt_client,
+            peer_receiver,
+            peer_senders,
+        })
+    }
+}
+
+#[async_trait]
+impl MessageBox for MqttMessageBox {
+    type Input = MqttActorMessage;
+    type Output = MqttActorMessage;
+
+    async fn recv(&mut self) -> Option<MqttActorMessage> {
+        tokio::select! {
```

Comment:
This is the exception that confirms the rule ;-) I would move this `tokio::select!` to the main loop of the actor and used a simple message box here.